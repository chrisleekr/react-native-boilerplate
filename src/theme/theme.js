import {Dimensions} from 'react-native';
import {colors} from './colors';
const {width, height} = Dimensions.get('window');

// Themes:- Can alter values here. Can only be consumed through Context (see useTheme.js file)
const themes = {
  light: {
    name: 'light',
    barStyle: 'light-content',
    color: '#000000',
    labelColor: '#525E75',
    invertColor: '#ffffff',
    layoutBg: '#f0f0f0',
    headerBg: '#f6f7f8',
    cardBg: '#e4dfa7',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#B00020',
    buttonActiveColor: '#6200EE',
    buttonDisabledColor: 'rgba(0, 0, 0, 0.26)',
    card: {
      shadowColor: '#000000',
      borderColor: '#B00020'
    },
    ...colors
  },
  dark: {
    name: 'dark',
    barStyle: 'dark-content',
    color: '#ffffff',
    labelColor: '#F1DDBF',
    invertColor: '#000000',
    layoutBg: '#121212',
    headerBg: '#1e1e1e',
    cardBg: '#3f4045',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#ffffff',
    buttonActiveColor: '#BB86FC',
    buttonDisabledColor: 'rgba(255,255,255, 0.26)',
    card: {
      shadowColor: '#121212',
      borderColor: '#ffffff'
    },
    ...colors
  }
};

const window = {
  // app dimensions
  width,
  height
};

export {themes, window};
