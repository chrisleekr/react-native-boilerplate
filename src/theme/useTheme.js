/**
 * For Dark/Light Mode
 * @exports ThemeContext, ThemeProvider, useTheme
 */

import React, {useState, createContext, useContext} from 'react';
import {Appearance} from 'react-native';

// Import preconfigured themes from theme file
import {themes} from './theme';

// Context
export const ThemeContext = createContext(null);

// Provider to be used in index/App/or top of any parent
export const ThemeProvider = ({children}) => {
  const colorScheme = Appearance.getColorScheme();

  const [theme, setTheme] = useState(themes[colorScheme]);
  return (
    <ThemeContext.Provider value={{theme, setTheme}}>
      {children}
    </ThemeContext.Provider>
  );
};

// useTheme hook for easy access to theme and setTheme
export const useTheme = () => {
  const {theme, setTheme} = useContext(ThemeContext);

  const toggleTheme = v => {
    setTheme(v ? themes.dark : themes.light);
  };

  return {theme, toggleTheme};
};
