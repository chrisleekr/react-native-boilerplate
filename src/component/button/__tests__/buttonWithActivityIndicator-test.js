import 'react-native';
import React from 'react';

import {cleanup, render, fireEvent} from '@testing-library/react-native';

import {ThemeProvider} from '../../../theme/useTheme';

import ButtonWithActivityIndicator from '../buttonWithActivityIndicator';

let rendered;
let element;
let onPress;
let isLoading;
let disabled;

const renderComponment = () =>
  render(
    <ThemeProvider>
      <ButtonWithActivityIndicator
        onPress={onPress}
        buttonStyle={{
          fontSize: 13
        }}
        text={'My Text'}
        isLoading={isLoading}
        disabled={disabled}
      />
    </ThemeProvider>
  );

describe('ButtonWithActivityIndicator', () => {
  afterEach(cleanup);

  describe('no loading, no disabled', () => {
    beforeEach(() => {
      onPress = jest.fn();
      isLoading = false;
      disabled = false;

      rendered = renderComponment();
    });

    it('matches snapshot', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });

    describe('checks disabled', () => {
      beforeEach(() => {
        element = rendered.getByTestId('button');
      });

      it('sets as active', () => {
        expect(element.props.accessibilityState.disabled).toBeFalsy();
      });
    });

    describe('activity indicator', () => {
      beforeEach(() => {
        element = rendered.queryByTestId('activity-indicator');
      });

      it('does not render activity indicator', () => {
        expect(element).toBeNull();
      });
    });

    describe('onPress', () => {
      beforeEach(() => {
        fireEvent.press(rendered.getByTestId('button'));
      });

      it('triggers onPress', () => {
        expect(onPress).toHaveBeenCalled();
      });
    });

    describe('text', () => {
      beforeEach(() => {
        element = rendered.getByText('My Text');
      });

      it('rendered text', () => {
        expect(element).toBeTruthy();
      });
    });
  });

  describe('isLoading', () => {
    beforeEach(() => {
      onPress = jest.fn();
      isLoading = true;
      disabled = false;

      rendered = renderComponment();
    });

    it('matches snapshot', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });

    describe('checks disabled', () => {
      beforeEach(() => {
        element = rendered.getByTestId('button');
      });

      it('sets as disabled', () => {
        expect(element.props.accessibilityState.disabled).toBeTruthy();
      });
    });

    describe('shows activity indicator', () => {
      beforeEach(() => {
        element = rendered.getByTestId('activity-indicator');
      });

      it('rendered activity indicator', () => {
        expect(element).toBeTruthy();
      });
    });
  });

  describe('disabled', () => {
    beforeEach(() => {
      onPress = jest.fn();
      isLoading = false;
      disabled = true;

      rendered = renderComponment();
    });

    it('matches snapshot', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });

    describe('checks disabled', () => {
      beforeEach(() => {
        element = rendered.getByTestId('button');
      });

      it('sets as disabled', () => {
        expect(element.props.accessibilityState.disabled).toBeTruthy();
      });
    });

    describe('does not show activity indicator', () => {
      beforeEach(() => {
        element = rendered.queryByTestId('activity-indicator');
      });

      it('does not render activity indicator', () => {
        expect(element).toBeFalsy();
      });
    });
  });
});
