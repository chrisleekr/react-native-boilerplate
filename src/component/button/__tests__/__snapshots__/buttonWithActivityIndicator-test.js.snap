// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonWithActivityIndicator disabled matches snapshot 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0, 0.26)",
      "borderColor": "black",
      "borderWidth": 1,
      "flexDirection": "row",
      "fontSize": 13,
      "height": 44,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 10,
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
  testID="button"
>
  <Text
    style={
      Array [
        undefined,
        Object {
          "color": "#ffffff",
        },
      ]
    }
  >
    My Text
  </Text>
</View>
`;

exports[`ButtonWithActivityIndicator isLoading matches snapshot 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0, 0.26)",
      "borderColor": "black",
      "borderWidth": 1,
      "flexDirection": "row",
      "fontSize": 13,
      "height": 44,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 10,
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
  testID="button"
>
  <View
    style={
      Array [
        Object {
          "marginRight": 5,
        },
      ]
    }
    testID="activity-indicator"
  >
    <ActivityIndicator
      color="#ffffff"
    />
  </View>
  <Text
    style={
      Array [
        undefined,
        Object {
          "color": "#000000",
        },
      ]
    }
  >
    My Text
  </Text>
</View>
`;

exports[`ButtonWithActivityIndicator no loading, no disabled matches snapshot 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#6200EE",
      "borderColor": "black",
      "borderWidth": 1,
      "flexDirection": "row",
      "fontSize": 13,
      "height": 44,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 10,
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
  testID="button"
>
  <Text
    style={
      Array [
        undefined,
        Object {
          "color": "#ffffff",
        },
      ]
    }
  >
    My Text
  </Text>
</View>
`;
